### ISSUE TITLE
Accessibility hint for capture process should be exposed as property on image capture step

### ISSUE THREAD

**Issue by jwe-apple** [ResearchKit/ResearchKit#228]
The image capture step should probably include an accessibility hint. Since the description of what exactly to do is specific to the particular use case, we should expose this as a property on the image capture step.

**Comment by brucehappy** [comment]
The ``Localizable.strings`` has two strings that have existed since the epoch of the project on github, and that seem to be remnants of a never included Media step. 
```
/* Media step */
"MEDIA_BUTTON_HAVE_IMAGE" = "Tap to capture a different image";
"MEDIA_BUTTON" = "Tap to capture an image";
```
This raises the question, do we need two hints?   One for the state where an image is not captured, and one for when it is already captured?  If so, presumably we need allow the developer to specify both hints via properties.  And, would you like to repurpose the above strings, with different names, for the default hints of the image capture step when the developer does not specify any?

**Comment by brucehappy** [comment]
@jwe-apple Also would like to know if you believe the hint should be on the Capture/Recapture buttons, or if it should be on the preview.  Currently, I've got the hint on the preview (switching it between the two hints depending on the current mode), and I've also made the magic tap on the preview click either the capture or recapture button depending on which is active.

**Comment by YuanZhu-apple** [comment]
@brucehappy 
"MEDIA_BUTTON_HAVE_IMAGE" and "MEDIA_BUTTON" are not being used, we need to clean it up.   

Add a hint for the image template on the top of preview view, so voice over users can understand the aim of the image capture. 

Currently,  it only says "Capture preview, image". Ideally, it would say something like: "Capture preview, right hand, image".

**Comment by umerkhan-apple** [comment]
https://github.com/ResearchKit/ResearchKit/pull/305

**Comment by brucehappy** [comment]
I had worked on this issue, as indicated in my previous comments.  Apologies for not submitting a PR after the response from @YuanZhu-apple .  I do not wish to step on anyone's toes, but I can submit the PR now, if desired.

**Comment by umerkhan-apple** [comment]
@brucehappy I figured you were busy, so I went ahead and started working on the issue. I have updated the PR following your suggestions. Please compare the PR with your version and let me know of any differences or suggestions. Let me know if you feel your version is significantly different and would like me to close my PR?

**Comment by brucehappy** [comment]
Ok, just ping me by replying to the issue in future, if you don't mind.  It could save us from duplicating work.

I can help you get this PR into shape if you want.  As I had indicated above, my code allows for two separate hints, one for when the preview is displaying a live feed from the camera, and one for when it is is showing the captured image.  @YuanZhu-apple seemed to be suggesting in his comment that only one was necessary, though he never specifically addressed it.  I have submitted my code as PR #306.  Feel free to draw code from it, use it as is, or close it as @YuanZhu-apple and @umerkhan-apple see fit.

**Comment by brucehappy** [comment]
> @umerkhan-apple wrote in #306:
> @brucehappy I'm happy with this implementation. I will close my PR and we can follow up on this one. Just going back to @YuanZhu-apple comment from before, do we need two accessibility hints? The captured image may not necessarily always be what the user was instructed to capture. Also, should the "accessibility hint"/instructions be required if a template image is being used? Or should it be required regardless? Or not required at all?

I would suggest that the accessibility hint not be required at all.  If they place an instruction step in front of the image capture step, then the screen reader would have read that before they get to the capture step, which in my mind, makes the accessibility hint optional.  Thus I would suggest leaving it to the developer to determine whether they need the hint on their capture step or not.

As for having two hints, I have no problem changing the code to remove the ``recaptureAccessibilityHint`` if we feel that is the right approach.  My thought process was to indicate to the screen reader that the preview view is now showing a captured image, instead of a live preview.  That is, to indicate to the user that the capture has occurred.  Certainly the camera sound will be made when they capture, but my thought was to reinforce that with the hint.  I see that in your PR you changed the accessibility label depending on whether an image was captured or not, which addresses this same concern.

Perhaps we should follow your approach and have a single accessibility hint, and two standard accessibility labels which change depending on current mode.  The first is "Capture preview", and the other is "Captured image", for instance.  In that case, we could include the full instructions, including about recapture, in the single accessibility hint.  So, it could say "Capture preview, image", or "Captured image, image", and then the hint, "Align your hand with the template and press the capture button.  Press the recapture button to begin the process again."

Thoughts?

**Comment by umerkhan-apple** [comment]
I agree with your decision on the requirement of accessibility hint. We should leave it to the developer to determine. For the 'recaptureAccessibilityHint', I think we should remove it and have two standard accessibility labels.

**Comment by brucehappy** [comment]
I reviewed the documentation on accessibility after reading your response, and thinking back over this.  Please look at the documentation for accessibilityHint here:
https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIAccessibility_Protocol/index.html#//apple_ref/occ/instp/NSObject/accessibilityHint

> An accessibility hint helps users understand what will happen when they perform an action on the accessibility element when that result is not obvious from the accessibility label.  For example, if you provide an Add button in your application, the button’s accessibility label helps users understand that tapping the button adds values in the application. If, on the other hand, your application allows users to play a song by tapping its title in a list of song titles, the accessibility label for the list row does not tell users this. To help an assistive application provide this information to users with disabilities, an appropriate hint for the list row would be “Plays the song.”

Although I have added an accessibility magic tap on the preview view, there is no action that can be performed on the preview view itself using the standard single finger double tap.  Instead, you perform the action using the "Capture Image" or "Recapture Image" buttons.  And in both of those cases, the button text and accessibility label seem sufficient to explain what the action is when they are pressed.  I know that @jwe-apple said:
>include an accessibility hint.....the description of what exactly to do is specific to the particular use case

But using the accessibility hint on the preview to explain "what to do", and not "what happens" seems contrary to the description of what an accessibility hint is for in the documentation.

I will make the changes discussed and update the PR, but I wanted to see if @YuanZhu-apple , @umerkhan-apple , and @jwe-apple could discuss my concerns internally given what the docs on accessibility hint say about its use.

**Comment by brucehappy** [comment]
I have implemented what was discussed.  Please note that Voice Over seems to remove duplicate words found in the ``accessibilityLabel`` and ``accessibilityHint``, so instead of saying "Captured image, image", it just says "Captured image" when the preview is showing a captured image.

Please let me know your thoughts about my above concerns on the use of the accessibilityHint.

**Comment by umerkhan-apple** [comment]
@brucehappy After internal discussion, we have come to the conclusion that it is best to include an accessibilityHint on the capture preview, but refactor it as accessibilityInstructions. This way it can be an implementation decision to add instructions on the capture preview or to include them as an accessibilityHint on the "Capture Image" button.

To avoid Voice Over removing duplicate words, we can either leave it as is or refactor "Captured image" as "Captured photo".

**Comment by brucehappy** [comment]
Ok, so you want the property called ``accessibilityHint`` in ``ORKImageCaptureStep`` to be renamed to ``accessibilityInstructions``.  Correct?

How are you proposing that people add an accessibilityHint to the "Capture Image" button?  Through an override on the ViewController, or some other means?

Or perhaps I've misunderstood and you want two properties on ``ORKImageCaptureStep``, one called ``accessibilityHint`` that is used on the "Capture Image" button, and another property called ``accessibilityInstructions`` that is used on the capture preview?

**Comment by umerkhan-apple** [comment]
The last option with ``accessibilityHint`` on "Capture Image" button and ``accessibilityInstructions`` on capture preview.

**Comment by brucehappy** [comment]
I have updated #306 with the new property and the renaming of the existing property.  I left the text used for the preview's accessibilityLabel alone.

**Comment by brucehappy** [comment]
PR updated with latest feedback.  Updated the properties to copy, added more documentation of the `accessibilityInstructions` property with example, and updated the text used for the `accessibilityInstructions` in ORKTest to be more realistic for assisting a visually-impaired user.

**Comment by YuanZhu-apple** [comment]
Resolved in #306
